pkg/quota/image/usage.go

```diff
@@ -10,139 +13,400 @@ import (
+			return imageapi.DockerImageReference{}, fmt.Errorf("%q is not currently pointing to an image, cannot use it as the source of a tag", objRef.Name)
+		}
+		return imageapi.ParseDockerImageReference(event.DockerImageReference)
+
+	case "DockerImage":
+		managedByOS, ref := imageReferenceBelongsToInternalRegistry(objRef.Name)
+		if !managedByOS {
+			return imageapi.DockerImageReference{}, fmt.Errorf("DockerImage %s does not belong to internal registry", objRef.Name)
 		}
+		return ref, nil
 	}
 
-	return
+	return imageapi.DockerImageReference{}, fmt.Errorf("unsupported object reference kind %s", objRef.Kind)
+}
+
+// getImageStream gets an image stream object from etcd and caches the result for the following queries.
+func (c *GenericImageStreamUsageComputer) getImageStream(namespace, name string) (*imageapi.ImageStream, error) {
+	key := fmt.Sprintf("%s/%s", namespace, name)
+	if is, exists := c.imageStreamCache[key]; exists {
+		return is, nil
+	}
+	is, err := c.osClient.ImageStreams(namespace).Get(name)
+	if err == nil {
+		c.imageStreamCache[key] = is
+	}
+	return is, err
+}
+
+// getImage gets image object from etcd and caches the result for the following queries.
+func (c *GenericImageStreamUsageComputer) getImage(name string) (*imageapi.Image, error) {
+	if image, exists := c.imageCache[name]; exists {
+		return image, nil
+	}
+	image, err := c.osClient.Images().Get(name)
+	if err == nil {
+		c.imageCache[name] = image
+	}
+	return image, err
+}
+
+// getImageStreamImage gets an image belonging to a given image stream object from etcd and caches the result for the following queries.
+func (c *GenericImageStreamUsageComputer) getImageStreamImage(namespace, name string) (*imageapi.Image, error) {
+	nameParts := strings.SplitN(name, "@", 2)
+	if len(nameParts) != 2 {
+		return nil, fmt.Errorf("failed to parse image stream image name %q, expected exactly one '@'", name)
+	}
+	image, exists := c.imageCache[nameParts[1]]
+	if exists {
+		return image, nil
+	}
+	isi, err := c.osClient.ImageStreamImages(namespace).Get(nameParts[0], nameParts[1])
+	if err == nil {
+		image = &isi.Image
+		c.imageCache[name] = image
+	}
+	return image, err
+}
+
+// listImageStreams returns a list of image streams of the given namespace and caches them for later access.
+func (c *GenericImageStreamUsageComputer) listImageStreams(namespace string) (*imageapi.ImageStreamList, error) {
+	iss, err := c.osClient.ImageStreams(namespace).List(kapi.ListOptions{})
+	if err == nil {
+		for _, is := range iss.Items {
+			c.imageStreamCache[fmt.Sprintf("%s/%s", namespace, is.Name)] = &is
+		}
+	}
+	return iss, err
+}
+
+// cacheInternalRegistryName caches registry name of the given docker image reference of an image stored in an
+// internal registry.
+func cacheInternalRegistryName(dockerImageReference string) {
+	ref, err := imageapi.ParseDockerImageReference(dockerImageReference)
+	if err == nil && len(ref.Registry) > 0 {
+		internalRegistryNames.Insert(ref.Registry)
```

Comment:
The rough plan here is that all "internal" images will be represented by a
stub value, and it should be clear from the API value that the field is
pointing to an image inside of OpenShift (or give the client the tools to
handle that).

On Mon, Mar 21, 2016 at 6:14 AM, Michal Minar <notifications@github.com>
wrote:

> In pkg/quota/image/usage.go
> <https://github.com/openshift/origin/pull/7930#discussion_r56801142>:
>
> > +func (c *GenericImageStreamUsageComputer) listImageStreams(namespace string) (*imageapi.ImageStreamList, error) {
> > +	iss, err := c.osClient.ImageStreams(namespace).List(kapi.ListOptions{})
> > +	if err == nil {
> > +		for _, is := range iss.Items {
> > +			c.imageStreamCache[fmt.Sprintf("%s/%s", namespace, is.Name)] = &is
> > +		}
> > +	}
> > +	return iss, err
> > +}
> > +
> > +// cacheInternalRegistryName caches registry name of the given docker image reference of an image stored in an
> > +// internal registry.
> > +func cacheInternalRegistryName(dockerImageReference string) {
> > +	ref, err := imageapi.ParseDockerImageReference(dockerImageReference)
> > +	if err == nil && len(ref.Registry) > 0 {
> > +		internalRegistryNames.Insert(ref.Registry)
>
> Shall we start storing registry urls in etcd? This is not the first place
> I needed to somehow fetch registry url. It always ends up with parsing some
> DockerImageReference. The problem with these references is that they may
> get outdated.
>
> In the follow up, I will move the cache to the image registry (which will
> make it effectively duplicated). But having it in one place easily
> accessible from anywhere would be really nice.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7930/files/61b56fa4ebd754527d69bc308ed60c75e3d14acc#r56801142>
>